### PR TITLE
Proposal for copyImageBitmapToTexture

### PR DESCRIPTION
**Pull request by kainino0x**
[Rendered](https://github.com/kainino0x/gpuweb/blob/imagebitmap/design/ImageBitmap.md)

Designed/written in collaboration with @kenrussell.

### CODE REVIEW

**Comment by Kangz** [comment]
Thanks for putting this together, the proposal looks good. Have we also investigated the tradeoffs for an unbuffered alternative?

**Comment by kainino0x** [comment]
We didn't consider it. I don't think it makes much of a difference - either way, the operation occurs atomically at the unbuffered operation (be it texture.copyFromImageBitmap or queue.submit), and would have the same semantics (I think).

File: design/ImageBitmap.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# WebGPU + ImageBitmap
+
+```webidl
+dictionary ImageBitmapCopyView {
+    ImageBitmap imageBitmap;
+    GPUOrigin3D origin;
```
Comment (kainino0x) [inline design/ImageBitmap.md]: JG: maybe change this origin to be 2D

**Comment by jdashg** [comment]
It would be nice to try to expand this to actual 3d sliced uploads later, but let's get this in first.

File: design/ImageBitmap.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+partial interface GPUCommandEncoder {
+    void copyImageBitmapToTexture(
+        ImageBitmapCopyView source,
+        GPUTextureCopyView destination,
+        GPUExtent3D copySize);
+};
+```
+
+`copyImageBitmapToTexture` encodes a copy from a source sub-rectangle of an `ImageBitmap` into a destination sub-resource of a `GPUTexture`.
+The `ImageBitmap` is referenced at the time the command buffer is submitted.
```
Comment (kainino0x) [inline design/ImageBitmap.md]: JG: If ImageBitmap is read-only, just say "if it's detached at submit, it's an error"

**Comment by kainino0x** [comment]
Done and rebased.

**Comment by Kangz** [comment]
I think this is ready to be landed, the group agreed to this PR a while ago.

File: design/ImageBitmapToTexture.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# WebGPU + ImageBitmap
+
+```webidl
+dictionary ImageBitmapCopyView {
+    ImageBitmap imageBitmap;
+    GPUOrigin3D origin;
```
Comment (frzi) [inline design/ImageBitmapToTexture.md]: This still references GPUOrigin3D, unlike the updated IDL, referencing the new GPUOrigin2D. :)

Comment (kainino0x) [reply design/ImageBitmapToTexture.md]: Thanks for catching this!

**Comment by kainino0x** [comment]
From notes:

> ImageBitmap #266
> 
> - Concern that the ImageBitmap needs to be remembered between encode and submission. Discussion around read-only and weak handles.
> - ImageBitmap has to be not detached.
> - People like the proposal (but no consensus recorded)

**Comment by kainino0x** [comment]
The alternative is for this copy to be outside of a command buffer (queue/device timeline). I think this is somewhat hard to spec without multiqueue spec.

There was some contingent, IIRC, who wanted to have versions of some copies outside of buffers. We should/will revisit moving this out if we do that.

**Pull request merged**